modules/nixos/buildbot.nix

```
@@ -25,6 +25,16 @@ in
   sops.secrets.buildbot-nix-workers = { };
   sops.secrets.cachix-auth-token = { };
 
+  services.buildbot-nix.master.pullBased = {
+    pollInterval = 1800;
+    repositories = {
+      "astro/microvm.nix" = {
+        url = "https://github.com/qowoz/microvm.nix.git";
+        defaultBranch = "main";
+      };
```

```suggestion
      "astro/microvm.nix" = {
        url = "https://github.com/qowoz/microvm.nix.git";
        defaultBranch = "main";
        evalWorkerCount = 4;
      };
```
@Mic92 Could we make evalWorkerCount configurable per repo? This repo uses a lot of memory and I really don't want to have to reduce it just for one repo.

---

@zowoq sounds doable and useful. Please open an issue. We have most of the code needed for this to happen already there.

---

I found it was easy to add it as a config file option so for now I'll use that: https://github.com/nix-community/buildbot-nix/commit/2d039c3557a82d19e63186a22c9196fafed6e664

Unfortunately found other problems. Even with just 8 workers it still uses a lot of memory, too much considering that we other jobs running. Also very slow to eval, after 20 minutes it hasn't even finished aarch64-linux, it'll block jobs other repos. Not sure if we'd want change the current eval lock to something like a per repo eval lock and limit the buildbot instance to 2 concurrent evals.

---

How many virtual machines does this have?
@astro are you sharing nixpkgs instances in your microvm machines.
Maybe we also need to make configurable what it evaluates i.e. only a single architecture.

---

I can easily get the test matrix above 128 tests. It's all individual flake check entries but the nixpkgs are pinned through flake.lock. microvm.nix overlay only.

---

`hyperfine -r 5 'nix-eval-jobs --workers 1 --force-recurse --flake github:astro/microvm.nix#hydraJobs.x86_64-linux.crosvm-startup-shutdown'`
```
Benchmark 1: nix-eval-jobs --workers 1 --force-recurse --flake github:astro/microvm.nix#hydraJobs.x86_64-linux.crosvm-startup-shutdown
  Time (mean ± σ):     91.982 s ±  0.193 s    [User: 54.372 s, System: 25.974 s]
  Range (min … max):   91.669 s … 92.138 s    5 runs
```
`hyperfine -r 5 'nix-eval-jobs --workers 1 --force-recurse --flake github:astro/microvm.nix#hydraJobs.aarch64-linux.qemu-9pstore-overlay-shutdown-command'`
```
Benchmark 1: nix-eval-jobs --workers 1 --force-recurse --flake github:astro/microvm.nix#hydraJobs.aarch64-linux.qemu-9pstore-overlay-shutdown-command
  Time (mean ± σ):     92.361 s ±  0.502 s    [User: 54.808 s, System: 26.118 s]
  Range (min … max):   91.557 s … 92.925 s    5 runs

```

---

Seems I only pasted the benchmarks and forgot to include a comment.

I found that lots of the hydraJobs attributes took around 90 seconds each to eval. 

The combination of the number of tests, the test attributes are slow to eval and that they also use a significant amount memory is a problem. I need to limit the number of parallel evals to reduce memory usage, which then makes the overall eval time to slow to be practical.

Good news is that soon we'll be able to offer buildbot CI app to external repos so if we can address eval issues you could have PR tests without needing to move the repo into nix-community.

---

The flamegraph eval profiler than landed in 2.30 may help with making some progress on this.

https://discourse.nixos.org/t/nix-2-30-0-released/66449